src/Camelot.ViewModels/Implementations/MainWindow/FilePanels/FilesPanelViewModel.cs

```diff
@@ -438,4 +444,108 @@ private int GetInsertIndex(IFileSystemNodeViewModel newNodeViewModel)
 
     private bool CheckIfShouldShowNode(string nodePath) =>
         _specification?.IsSatisfiedBy(_nodeService.GetNode(nodePath)) ?? true;
-}
\ No newline at end of file
+
+    private int GetSelectedIndex()
+    {
+        if (SelectedFileSystemNodes.Count == 0)
+            return -1;
+        var oldSelected = SelectedFileSystemNodes.First();
+        var nodes = FileSystemNodes.ToList();
+        for (int i = 0; i < nodes.Count; i++)
+        {
+            var curr = nodes[i];
+            if (curr.FullPath == oldSelected.FullPath)
+            {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private IFileSystemNodeViewModel GetSelected()
+    {
+        int selectedIndex = GetSelectedIndex();
+        if (selectedIndex < 0)
+            return null;
+
+        var nodes = FileSystemNodes.ToList();
+        var selected = nodes[selectedIndex];
+        return selected;
+    }
+
+    private void SelectNodeEx(string newSelected, IFileSystemNodeViewModel oldSelected)
+    {
+        if (newSelected != null)
+        {
+            if (oldSelected != null)
+            {
+                UnselectNode(oldSelected.FullPath);
+            }
+            SelectNode(newSelected);
+        }
+    }
+
+    public void OnDataGridKeyDownCallback(Key key)
```

Comment:
ViewModels by definition should not have code that looks like code behind callback method. This method should be renamed to represent the business logic it contains